packages/edit-site/src/components/sync-state-with-url/use-sync-path-with-url.js

```diff
@@ -44,76 +50,79 @@ export default function useSyncPathWithURL() {
-				...currentUrlParams.current,
-				...newUrlParams,
-			};
-			currentUrlParams.current = updatedParams;
-			history.push( updatedParams );
-		}
 
-		if ( navigatorParams?.postType && navigatorParams?.postId ) {
-			updateUrlParams( {
-				postType: navigatorParams?.postType,
-				postId: navigatorParams?.postId,
-				path: undefined,
-			} );
-		} else if (
-			navigatorLocation.path.startsWith( '/page/' ) &&
-			navigatorParams?.postId
-		) {
-			updateUrlParams( {
-				postType: 'page',
-				postId: navigatorParams?.postId,
-				path: undefined,
-			} );
-		} else if ( navigatorLocation.path === '/patterns' ) {
-			updateUrlParams( {
-				postType: undefined,
-				postId: undefined,
-				canvas: undefined,
-				path: navigatorLocation.path,
-			} );
-		} else {
-			updateUrlParams( {
-				postType: undefined,
-				postId: undefined,
-				categoryType: undefined,
-				categoryId: undefined,
-				path:
-					navigatorLocation.path === '/'
-						? undefined
-						: navigatorLocation.path,
-			} );
-		}
-	}, [ navigatorLocation?.path, navigatorParams, history ] );
+			function updateUrlParams( newUrlParams ) {
+				if ( isSubset( newUrlParams, urlParams ) ) {
+					return;
+				}
+				const updatedParams = {
+					...urlParams,
+					...newUrlParams,
+				};
+				history.push( updatedParams );
+			}
 
-	useEffect( () => {
-		currentUrlParams.current = urlParams;
-		const path = getPathFromURL( urlParams );
-		if ( currentPath.current !== path ) {
-			currentPath.current = path;
-			goTo( path );
-		}
-	}, [ urlParams, goTo ] );
+			if ( navigatorParams?.postType && navigatorParams?.postId ) {
+				updateUrlParams( {
+					postType: navigatorParams?.postType,
+					postId: navigatorParams?.postId,
+					path: undefined,
+				} );
+			} else if (
+				navigatorLocation.path.startsWith( '/page/' ) &&
+				navigatorParams?.postId
+			) {
+				updateUrlParams( {
+					postType: 'page',
+					postId: navigatorParams?.postId,
+					path: undefined,
+				} );
+			} else if ( navigatorLocation.path === '/patterns' ) {
+				updateUrlParams( {
+					postType: undefined,
+					postId: undefined,
+					canvas: undefined,
+					path: navigatorLocation.path,
+				} );
+			} else {
+				updateUrlParams( {
+					postType: undefined,
+					postId: undefined,
+					categoryType: undefined,
+					categoryId: undefined,
+					path:
+						navigatorLocation.path === '/'
+							? undefined
+							: navigatorLocation.path,
+				} );
+			}
+		},
+		// Trigger only when navigator changes to prevent infinite loops.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ navigatorLocation.path, navigatorParams ]
+	);
+
+	// Update navigator path when the URL params change.
+	useEffect(
+		() => {
+			const path = getPathFromURL( urlParams );
+			if ( navigatorLocation.path !== path ) {
+				goTo( path );
+			}
+		},
+		// Trigger only when URL changes to prevent infinite loops.
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ urlParams ]
+	);
```

Comment:
Can you clarify the changes in this hook. The hook is supposed to be independent of the navigation, just synchronizes changes, so I'm not sure how it's related to this PR

Reply:
Yep so the key change, and all that this PR needs at minimum, is changing this:

https://github.com/WordPress/gutenberg/blob/181b49cde2b5ddbe5d0ca164029c5ab98ee13580/packages/edit-site/src/components/sync-state-with-url/use-sync-path-with-url.js#L113-L117

To this:

https://github.com/WordPress/gutenberg/blob/e83e1afc0a55a4531ee3012bf824611626a35e01/packages/edit-site/src/components/sync-state-with-url/use-sync-path-with-url.js#L117-L120

The former is buggy. When something calls `navigator.goTo`, e.g. when you click _Pages_, the result is:

1. The navigator path changes to `/page`
2. The first `useEffect` fires (because the navigator path changed) and calls `history.push` to change the URL to `?path=%2Fpage`
3. The second `useEffect` fires (because the URL params changed) and calls `navigator.goTo` to navigate to `/page`

(3) is extraneous. We don't need to call `navigator.goTo( '/page' )` as the navigator location is already `/page`. Doing this results in two history items within the navigator and means that the user has to click back _twice_ to return to the root page.

The simplest way to fix this would be to update `currentPath.current` in the first `useEffect`, but I thought I'd go further and attempt to simplify `useSyncPathWithURL` to not require the `currentPath` and `currentUrlParams` refs at all. They duplicate state that is already in the `useHistory` and `useNavigator` hooks. Having the refs makes the hook more difficult to understand and makes it easy for bugs of this nature to slip by.